Cache cleanup

akirapop opened the pull request:
I have modified the original code to now use the factored out code that was in the rpm_cleanup branch (and now is in master). A dry run gives the correct resulsts -- all cache directories with non-existent or "stale" (no commits within 2 weeks) branches are setup to be removed.

akirapop committed:
First commits.

akirapop committed:
Second commit.

akirapop committed:
First trial script

akirapop committed:
Edited Jenkinsfile

akirapop committed:
Bug fixing.

akirapop committed:
Bug found, I think.

akirapop committed:
Refactoring.

akirapop committed:
Merge branch 'master' into cache_cleanup
I made lots of changes here, incorporating factorized code that is also used in
rpm repository cleanups.

akirapop committed:
Tested the code via a dry run. Things seem ok, so I have "activated" the code
via removing the '#' on the "rm -rf ${dir}" line. This code will be 'live', then.

Metallion commented:
Would it be possible to put it under the garbage collection job before merge? I think it would make the CI layout more clean.

For running them in parallel maybe this can help: https://github.com/jenkinsci/pipeline-examples/blob/master/docs/BEST_PRACTICES.md#parallelism

If it looks like it will take too much time, we could merge already and fix that afterwards.

akirapop committed:
Quick bug fix here.

akirapop committed:
I have added code to allow for all garabage collections jobs to
be run under the *same* jenkins job. This is done via "parallel"
in the Jenkinsfile.

akirapop committed:
I updated this file so that it matches the one in master.

akirapop commented:
I have re-structured things now so that the rpm_cleanup and cache_cleanup code are run in parallel as part of one jenkins job (the garbage_collection job). You can take a look at the recent output. It is probably better to look at the "Console Output", as it is clearer to see that the jobs were run in parallel.

Please hold on merging, though: if you look at the "live" run from this morning -- it failed! The reason: there are permission problems.

Metallion commented:
Code looks good to me. Also looks like the permission problem is solves so is this ok to merge?

akirapop committed:
I've done some cleaning up, removing no longer needed Jenkinsfiles.

akirapop committed:
Merge branch 'master' into cache_cleanup

akirapop committed:
Removed comment from "rm -rf" line. We are now 'live' again.

akirapop committed:
Bugchecking here. Trying to clarify who 'I' am and what my groups are.

akirapop committed:
I am experimenting: can I change the jenkins groups and then execute the scripts?

akirapop committed:
Restarted jenkins slave. We will see if the envt. is now ok.

akirapop committed:
I am still trying to find out why the groups for user jenkins are not
what it needs to be.

akirapop committed:
Some of the directories to be removed in the cache-cleaning section
are owned by root. As a temporary workaround, I have added code to
check to see if a given directory is owned by root. If it is, the code
will not try to remove it (the directory cannot be removed)!

akirapop committed:
Silly bugfix here.

akirapop committed:
In order to allow the jenkins user to remove no longer needed
rpm repos., the yumrepo umask must be changed. I've added a line
to the code that creates the directory+rpm repo that changes the
umask to g+w.

akirapop committed:
Made another change to allow change in group of yumrepo.

akirapop committed:
Bugfix here. Gave the wrong environment var. name!

akirapop commented:
Things seem to be in order now -- both the cache cleanup and the rpm cleanup section are able to remove "stale" directories. *BUT:  Note that the problem I mentioned yesterday with some caches now being created by "root". These directories cannot be removed. I've added code that checks to see if the directory owner is root. If so, no rm -rf ...

One note: to change permissions, a "chgrp -R repoci ${RPM_ABSOLUTE}" & "chmod -r g+w ${RPM_ABSOLUTE}" were run.

akirapop committed:
Added code to Jenkinsfile.garbage_collection to account for the
not-yet merged docker cleanup code.

Metallion commented on ci/garbage_collection/cache_cleanups/cache_garbage_collection.sh:
```diff
@@ -0,0 +1,29 @@
+time_limit=14     ## Days. Set this to give the "deadline". All
+                  ## branches older than this will be removed.
+
+cache_location_dir=/data2/openvdc-ci/branches
```
Needs to be changed to `/data` because of the move to dh112. Could be a good idea to read this from the multibox configuration but just hard coding will work for now imo.

Metallion commented on ci/garbage_collection/repo_cleanups/rpm_garbage_collection.sh:
```diff
@@ -25,7 +25,6 @@ cd ${rpm_base_dir}/master
 nrepos=$(ls -1 . | wc -l)
 if [[ ${nrepos} -lt 2 ]]; then
    echo "Something is wrong. The master directory contains one or less repos. Quitting."
-   exit 1
```
This removal makes the above message wrong. :)

Metallion commented on ci/garbage_collection/cache_cleanups/cache_garbage_collection.sh:
```diff
@@ -0,0 +1,29 @@
+
+   group_owner=$(stat -c %G ${cache_location_dir}/${directory})
+   if [[ "${group_owner}" = "root" ]]; then
+       echo "${directory} is owned by root. Cannot remove!"
```
The problem is that currently the acceptance test's cache is always owned by root. This code is ok like this but the multibox code should change the cache images group settings so this script can remove them.

akirapop committed:
Made corrections/edits as per comments made on the pull request.

Metallion commented:
There's still the problem where the test cache is always owned by root and thus will never be deleted by this. It would be good to fix that in the same PR. I'll also have a look to see exactly how we should do those settings. The things is that the cache is currently made from a privileged docker container and that's why they're owned by root. A quick hack would probably to do a chown -R or chgrp -R at the end of the acceptance test.

akirapop committed:
Added code here to change the group of the ${CACHE_DIR} to "jenkins" from "root". This
is necessary for the cache cleanup code.

akirapop committed:
Fix of simple mis-spelling bug. Also, modified Jenkinsfile for the
garbage cleanup, as the docker image cleanup is not yet a part of this
branch.

akirapop committed:
Changing from using chgrp to chown: I think it's best to just change everything
here from user "root" to user "jenkins".
